core/src/main/scala/org/http4s/Media.scala

```diff
@@ -8,23 +8,32 @@ package org.http4s
 
 import cats.MonadError
 import cats.implicits._
-import fs2.Stream
+import fs2.{RaiseThrowable, Stream}
 import fs2.text.utf8Decode
 import org.http4s.headers._
-import org.http4s.util.decode
 
 trait Media[F[_]] {
   def body: EntityBody[F]
   def headers: Headers
   def covary[F2[x] >: F[x]]: Media[F2]
 
+  @deprecated("Can go into an infinite loop for charsets other than UTF-8. Replaced by bodyText", "0.21.5")
   final def bodyAsText(implicit defaultCharset: Charset = DefaultCharset): Stream[F, String] =
     charset.getOrElse(defaultCharset) match {
       case Charset.`UTF-8` =>
         // suspect this one is more efficient, though this is superstition
         body.through(utf8Decode)
       case cs =>
-        body.through(decode(cs))
+        body.through(util.decode(cs))
+    }
+
+  final def bodyText(implicit RT: RaiseThrowable[F], defaultCharset: Charset = DefaultCharset): Stream[F, String] =
```

Comment:
I don't love this implicit charset pattern that we have.  Maybe it's time to rip that out.